OverlayPlugin.Core/MemoryProcessors/AtkGui/Unmanaged/AddonPartyList.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System.Runtime.InteropServices;
+
+namespace RainbowMage.OverlayPlugin.MemoryProcessors.AtkGui.Unmanaged
+{
+    // https://github.com/goatcorp/FFXIVClientStructs/blob/main/FFXIVClientStructs/FFXIV/Client/UI/AddonPartyList.cs
+    [StructLayout(LayoutKind.Explicit, Size = 0x1420)]
+    public unsafe struct AddonPartyList
```

Comment:
This seems like the kind of thing that will need to be versioned, e.g. https://github.com/goatcorp/FFXIVClientStructs/commit/0a183151fbd2de08c5227c658150df2aa717a69e

EDIT: and same comment about most everything else in this directory.

Reply:
I have a complete rework of the underlying structs in the next PR which will allow them to be versioned properly, I'm just working out an issue related to your other comment on PartyListMemory62.cs

Reply:
The latest force-pushed commit has the entire struct setup from FFXIVClientStructs, trimmed to only fields. This can be trimmed further, but as it stands it's a question of "do we pull in the entire structure now to reduce commit size in the future".

If you'd prefer that this PR only have the structs required for `AddonPartyList` be included, that can be done instead. I just wanted to get an idea of the scope of the entire structure.

Reply:
This is...a lot of things.

I guess my concerns are:
* are we going to use all of these things? I have a weak preference for pulling in what we need when somebody needs it, because ...
* how are these going to be updated on a regular basis when releases bump them?
* how are these going to be all found for non-global versions?

The more memory offsets exist, the more work it is.  It's already a bunch of work to chase down the ~4 offsets for targeting for all versions, let alone thousands of offsets.

Also, what's the versioning story here? Are we going to have a separate sibling directory next to global?

Reply:
>     * are we going to use all of these things? I have a weak preference for pulling in what we need when somebody needs it, because ...
I haven't gone through and manually trimmed the stuff that's unused, but ultimately if we want to e.g. offer a full representation of `AddonPartyList` then that's going to require a large chunk of the files.

>     * how are these going to be updated on a regular basis when releases bump them?
Preferably, there would be an automated way to get these pulled in, I'm just not familiar enough with the C#/.NET side of the equation to figure out how to do so. I'd like to see e.g. the `build.ps1` script pull down the specific commits for each client version, perform a mass namespace rename in the pulled down project, and compile those into dlls that we can target directly. But I'm not sure exactly how to go about that process or if it's even possible because for example the FFXIVClientStructs project uses C# language version 9 syntax, and I'm not sure if we can target a compile that'll work against .NET 4.8

>     * how are these going to be all found for non-global versions?
For the actual struct layouts, presumably those are available as well since Dalamud must get them from somewhere, I'm just not sure exactly where. We may need to ask on goat's discord if they have a more direct answer.

> The more memory offsets exist, the more work it is. It's already a bunch of work to chase down the ~4 offsets for targeting for all versions, let alone thousands of offsets.
Only sigs for top-level objects need to be found. Every single `Addon` entry from the PR's initial comment can be found directly from the single sig added in this PR.

Additionally, those addresses are available directly from the FFXIVClientStructs project.
The sig I mentioned above is: 
https://github.com/aers/FFXIVClientStructs/blob/1ff6e4e87fcda058970f82b307233fde36d38a88/FFXIVClientStructs/FFXIV/Component/GUI/AtkStage.cs#L18-L19
And the static offset is:
https://github.com/aers/FFXIVClientStructs/blob/1ff6e4e87fcda058970f82b307233fde36d38a88/ida/sigdata.json#L51

> Also, what's the versioning story here? Are we going to have a separate sibling directory next to global?
Given that the structure and offsets can vary per region even if they're on the same client version (see #93) it's better to separate per region instead of per game version, I think.

Reply:
Attempt number three for pulling in the structs themselves in the latest commit (https://github.com/OverlayPlugin/OverlayPlugin/pull/130/commits/25aeefd4e664d064c38d71309482c52ff8d16e3c).

Basically, pull in a snapshot of the FFXIVClientStructs code, clean it up to compile against netstandard2.0, compile it, and then link against it.

Problem is:

```
\OverlayPlugin.Core\MemoryProcessors\PartyList\PartyListMemoryManager.cs(11,37)
: error CS8370: Feature 'unmanaged constructed types' is not available in C# 7.3. Please use language version 8.0 or gr
eater.
```

Reply:
Alright, this latest commit resolves the issue by building the structs into the application. This should be fully working and the event handler should return good data.

It's probably not going to survive bumping FFXIVClientStructs versions very well, but given that we're pinning to a given commit hash that's probably okay? Feedback on how better to handle this code transform process is welcomed.

Also, we can get the offset directly from the `ida/data.yml` file as well if that's something you think is worthwhile. See:
https://github.com/aers/FFXIVClientStructs/blob/8506108f3772ddcb608cf7cf0612ce6cb2b9bbf7/ida/data.yml#L2276-L2278

An additional benefit of pulling in the FFXIVClientStructs code is that we can replace e.g. the unmanaged struct for CombatantMemory with their equivalent `BattleChara` if they have all the fields mapped (and if they don't we can probably PR the missing fields up to them), reducing maintainer burden. It's probably the same with most of the other memory we're reading as well, though I haven't investigated anything else. Probably the appropriate sigs for these are also in `data.yml`, though again I haven't done much investigation.

Reply:
I think overall this is definitely the best patch yet, from my perspective.  Do you feel like there's more stuff to do or to answer here, or do you feel like this is where you'd want it to be before landing it?

I'm not sure what to do about the code transform process.  It sure looks like a lot, but it's certainly better than doing it manually.  I agree with your concern about how well it will survive being bumped.  I agree that having it be pinned will prevent bugs, but more a question of how much work will occur when it changes.  I think that's something we can figure out as we go.

Re: offsets.  I'm a little less sure? I think my previous concern re: how often this changes is still there.  The offset seems like it might change on every release, but structs change less frequently.  Ideally to me is some sort of signature that can live across releases if possible so that there's not "day 1 pressure" to update this and release OverlayPlugin so that party lists are working again.  Obviously if the structs change, we'll have to wait on that.

Replacing manual structs with FFXIVClientStructs seems like a nice change too.

I'm not sure I understand enough about C# to know what the implications of trying to bump our language version (or if we even can given other limitations like what SDK we're on?).  Is that something we should consider so we can use more of the files directly?

Reply:
> Do you feel like there's more stuff to do or to answer here, or do you feel like this is where you'd want it to be before landing it?

The current `getPartyList` event handler is still placeholder and my previous comment is still a thought I'd like feedback on, "It may even make sense to put this in its own ClientStructEventSource.cs file with an additional getFullAddonSlow handler that takes a string argument."

> Re: offsets. I'm a little less sure? I think my previous concern re: how often this changes is still there. The offset seems like it might change on every release, but structs change less frequently. Ideally to me is some sort of signature that can live across releases if possible so that there's not "day 1 pressure" to update this and release OverlayPlugin so that party lists are working again.

The data.yml file is autogenerated via `ida/ffxiv_sigmaker.py` but my extremely, extremely basic knowledge of python means I have no idea what it's actually doing to actually replicate it as a sigscan like we normally do at runtime.

> Obviously if the structs change, we'll have to wait on that.

Looks like 6.2 was the last time structs changed due to a patch, and everything relevant to us was updated by 1PM EDT.

 > I'm not sure I understand enough about C# to know what the implications of trying to bump our language version (or if we even can given other limitations like what SDK we're on?). Is that something we should consider so we can use more of the files directly?

The main limitation is that we're still on the old project format, we'd need to land #82 first. Then we could technically change the target version to .NET standard 2.0, which should be loadable by .NET 4.8 apps assuming we don't use any new libraries (it's technically an unsupported configuration, loading .NET standard 2.0 libraries in a .NET 4.8 application, but it should work). But that would require some extensive testing, I'm not sure exactly what might break.

Reply:
Latest commit has the actual full change set, minus the issue with offsets. I think that in general `ManagedType` and the other new classes added in this PR are ready for review.

Given the type mangling and reflection going on in ManagedType I'd like to make a final pass on it to add some comments and documentation, but it's functionally code-complete now.